manabi/lock.py

```diff
@@ -94,3 +117,125 @@ def get_lock_list(self, path, *, include_root, include_children, token_only):
                 include_children=include_children,
                 token_only=token_only,
             )
+
+
+class PostgresDict(MutableMapping):
+    def __init__(self, connection: psycopg2.extensions.connection):
+        self._connection = connection
+
+    @staticmethod
+    def decode_lock(lock):
+        # Encoding is not needed since a string is also valid for wsgidav
+        if "owner" in lock:
+            owner = lock["owner"]
+            if isinstance(owner, bytes):
+                lock["owner"] = lock["owner"].decode("UTF-8")
+
+    def cursor(self) -> psycopg2.extensions.cursor:
+        return self._connection.cursor()
+
+    def cleanup(self):
+        cursor = self.cursor()
+        cursor.execute("DELETE FROM manabi_lock;")
+
+    def __len__(self):
+        cursor = self.cursor()
+        cursor.execute("SELECT count(*) FROM manabi_lock")
+        return int(cursor.fetchone())
+
+    def __iter__(self):
+        cursor = self.cursor()
+        cursor.execute("SELECT token FROM manabi_lock")
+        for token in cursor.fetchall():
+            yield token[0]
+
+    def __delitem__(self, token):
+        cursor = self.cursor()
+        cursor.execute("DELETE FROM manabi_lock WHERE token = %s", (str(token),))
+
+    def __contains__(self, token):
+        cursor = self.cursor()
+        cursor.execute("SELECT * FROM manabi_lock WHERE token = %s", (str(token),))
+        if cursor.fetchone() is None:
+            return False
+        else:
+            return True
+
+    def __setitem__(self, token, lock):
+        cursor = self.cursor()
+        self.decode_lock(lock)
+        json_lock = json.dumps(lock)
+        cursor.execute(
+            """
+                INSERT INTO manabi_lock(token, data) VALUES (%(token)s, %(data)s)
+                    ON CONFLICT(token) DO
+                    UPDATE SET data = %(data)s WHERE manabi_lock.token = %(token)s
+            """,
+            {
+                "token": token,
+                "data": json_lock,
+            },
+        )
+
+    def __getitem__(self, token):
+        cursor = self.cursor()
+        cursor.execute("SELECT data FROM manabi_lock WHERE token = %s", (str(token),))
+        lock = cursor.fetchone()
+        if lock is None:
+            raise KeyError(f"{token} not found")
+        lock = lock[0]
+        return lock
+
+
+class PostgresReadWriteLock:
+    def __init__(self, connection):
+        connection.commit()
+        connection.autocommit = False
+        connection.set_session(
+            isolation_level=psycopg2.extensions.ISOLATION_LEVEL_SERIALIZABLE
+        )
```

Comment:
I think this connection initialisation stuff would be better off at the place where the connection is made. Otherwise, you'll run into weird issues if you're using a connection from elsewhere, and the other code behaves differently because a postgres lock was used

Reply:
Yes I moved that